File: app/src/main/java/com/njlabs/showjava/utils/ZipUtils.java
package com.njlabs.showjava.utils;

import android.os.Environment;

import com.njlabs.showjava.utils.logging.Ln;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

public class ZipUtils {

    public static File zipDir(File dir, String packageId){

        File zipIntoDir = new File(Environment.getExternalStorageDirectory() + "/ShowJava/archives/");

        if(!zipIntoDir.exists() || !zipIntoDir.isDirectory()){
            zipIntoDir.mkdirs();
        }

        File zipFile = new File(zipIntoDir, packageId+".zip");

        if(zipFile.exists()) {
            zipFile.delete();
        }

        try {

            ZipOutputStream zip;
            FileOutputStream fileWriter;
            fileWriter = new FileOutputStream(zipFile);
            zip = new ZipOutputStream(fileWriter);
            addFolderToZip("", dir.toString(), zip);
            zip.flush();
            zip.close();

        } catch (Exception e) {
            Ln.e(e);
        }

        return zipFile;
    }

    static private void addFileToZip(String path, String srcFile, ZipOutputStream zip) throws Exception {
        File folder = new File(srcFile);
        if (folder.isDirectory()) {
            addFolderToZip(path, srcFile, zip);
        } else {
            byte[] buf = new byte[1024];
            int len;
            FileInputStream in = new FileInputStream(srcFile);
            zip.putNextEntry(new ZipEntry(path + "/" + folder.getName()));
            while ((len = in.read(buf)) > 0) {
                zip.write(buf, 0, len);
            }
        }
    }

    static private void addFolderToZip(String path, String srcFolder, ZipOutputStream zip) throws Exception {
        File folder = new File(srcFolder);
        for (String fileName : folder.list()) {
            if (path.equals("")) {
                addFileToZip(folder.getName(), srcFolder + "/" + fileName, zip);
            } else {
                addFileToZip(path + "/" + folder.getName(), srcFolder + "/" + fileName, zip);
            }
        }
    }

    public static void unzip(File zipFile, File targetDirectory) throws IOException {
        unzip(zipFile, targetDirectory, System.out);
    }

    public static void unzip(File zipFile, File targetDirectory, PrintStream printStream) throws IOException {
        ZipInputStream zis = new ZipInputStream(new BufferedInputStream(new FileInputStream(zipFile)));
        try {
            ZipEntry ze;
            int count;
            byte[] buffer = new byte[8192];
            while ((ze = zis.getNextEntry()) != null) {
                File file = new File(targetDirectory, ze.getName());
                printStream.println(ze.getName());
                File dir = ze.isDirectory() ? file : file.getParentFile();
                if (!dir.isDirectory() && !dir.mkdirs()){
                    throw new FileNotFoundException("Failed to ensure directory: " +
                            dir.getAbsolutePath());
                }

                if (ze.isDirectory()) {
                    continue;
                }

                FileOutputStream fout = new FileOutputStream(file);
                try {
                    while ((count = zis.read(buffer)) != -1)
                        fout.write(buffer, 0, count);
                } finally {
                    fout.close();
                }
                long time = ze.getTime();
                if (time > 0) {
                    file.setLastModified(time);
                }
            }
        } finally {
            zis.close();
        }
    }
}

File: app/src/main/kotlin/com/njlabs/showjava/utils/ZipUtils.kt
/*
 * Show Java - A java/apk decompiler for android
 * Copyright (c) 2018 Niranjan Rajendran
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.njlabs.showjava.utils


import com.njlabs.showjava.utils.ktx.appStorage
import timber.log.Timber
import java.io.*
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

/**
 * [ZipUtils] to extract and compress directories
 */
object ZipUtils {

    fun zipDir(dir: File, packageId: String): File {
        val zipIntoDir = appStorage.resolve("archives")
        if (!zipIntoDir.exists() || !zipIntoDir.isDirectory) {
            zipIntoDir.mkdirs()
        }
        val zipFile = File(zipIntoDir, "$packageId.zip")
        if (zipFile.exists()) {
            zipFile.delete()
        }
        try {
            val zip: ZipOutputStream
            val fileWriter = FileOutputStream(zipFile)
            zip = ZipOutputStream(fileWriter)
            addFolderToZip("", dir.toString(), zip)
            zip.flush()
            zip.close()
        } catch (e: Exception) {
            Timber.e(e)
        }

        return zipFile
    }

    @Throws(Exception::class)
    private fun addFileToZip(path: String, srcFile: String, zip: ZipOutputStream) {
        val folder = File(srcFile)
        if (folder.isDirectory) {
            addFolderToZip(path, srcFile, zip)
        } else {
            val buffer = ByteArray(1024)
            val fileInputStream = FileInputStream(srcFile)
            zip.putNextEntry(ZipEntry(path + "/" + folder.name))

            while (true) {
                val len = fileInputStream.read(buffer)
                if (len <= 0) {
                    break
                }
                zip.write(buffer, 0, len)
            }
        }
    }

    @Throws(Exception::class)
    private fun addFolderToZip(path: String, srcFolder: String, zip: ZipOutputStream) {
        val folder = File(srcFolder)
        for (fileName in folder.list()) {
            if (path == "") {
                addFileToZip(folder.name, "$srcFolder/$fileName", zip)
            } else {
                addFileToZip(path + "/" + folder.name, "$srcFolder/$fileName", zip)
            }
        }
    }

    @Throws(IOException::class)
    @JvmOverloads
    fun unzip(zipFile: File, targetDirectory: File, printStream: PrintStream = System.out) {
        val zipInputStream = ZipInputStream(BufferedInputStream(FileInputStream(zipFile)))
        zipInputStream.use { zis ->
            val buffer = ByteArray(8192)
            while (true) {
                val zipEntry = zis.nextEntry ?: break
                val file = File(targetDirectory, zipEntry.name)
                printStream.println(zipEntry.name)
                val dir = if (zipEntry.isDirectory) file else file.parentFile
                if (!dir.isDirectory && !dir.mkdirs()) {
                    throw FileNotFoundException("Failed to ensure directory: " + dir.canonicalPath)
                }

                if (zipEntry.isDirectory) {
                    continue
                }

                val fileOutputStream = FileOutputStream(file)
                fileOutputStream.use { out ->
                    while (true) {
                        val count = zis.read(buffer)
                        if (count <= 0) {
                            break
                        }
                        out.write(buffer, 0, count)
                    }
                }
                val time = zipEntry.time
                if (time > 0) {
                    file.setLastModified(time)
                }
            }
        }
    }
}